src/TypeDefinition.js

```diff
@@ -54,42 +56,68 @@ export type NavigationRoute = {
   params?: NavigationParams,
 };
 
+export type NavigationStateRoute = {
+  /**
+   * React's key used by some navigators. No need to specify these manually,
+   * they will be defined by the router.
+   */
+  key: string,
+  /**
+   * For example 'Home'.
+   * This is used as a key in a route config when creating a navigator.
+   */
+  routeName: string,
+  /**
+   * Path is an advanced feature used for deep linking and on the web.
+   */
+  path?: string,
+  /**
+   * Params passed to this route when navigating to it,
+   * e.g. `{ car_id: 123 }` in a route that displays a car.
+   */
+  params?: NavigationParams,
 
-export type NavigationRouter = {
+
+  index: number,
+  routes: Array<NavigationRoute>,
+};
+
+
+export type NavigationRouter<State, Action, Options> = {
   /**
    * The reducer that outputs the new navigation state for a given action, with
    * an optional previous state. When the action is considered handled but the
    * state is unchanged, the output state is null.
    */
   getStateForAction: (
-    action: NavigationAction,
-    lastState: ?NavigationState,
-  ) => ?NavigationState,
+    action: Action,
+    lastState: ?State,
+  ) => ?State,
 
   /**
    * Maps a URI-like string to an action. This can be mapped to a state
    * using `getStateForAction`.
    */
-  getActionForPathAndParams: (path: string, params?: NavigationParams) => ?NavigationAction,
+  getActionForPathAndParams: (path: string, params?: NavigationParams) => ?Action,
 
-  getPathAndParamsForState: (state: NavigationState) => {path: string, params?: NavigationParams},
+  getPathAndParamsForState: (state: State) => {path: string, params?: NavigationParams},
 
   getComponentForRouteName: (routeName: string) => NavigationComponent,
 
-  getComponentForState: (state: NavigationState) => NavigationComponent,
+  getComponentForState: (state: State) => NavigationComponent,
 
   /**
-   * Gets the screen config for a given navigation screen prop.
+   * Gets the screen navigation options for a given screen.
    *
-   * For example, we could get the config for a 'Foo' screen when the
+   * For example, we could get the config for the 'Foo' screen when the
    * `navigation.state` is:
    *
    *  {routeName: 'Foo', key: '123'}
    */
-  getScreenConfig: (
-    navigation: NavigationScreenProp<NavigationRoute, NavigationAction>,
-    optionName: string,
-  ) => ?any, // todo, fix this any type to become a key of NavigationScreenConfig
+  getScreenOptions: (
+    navigation: NavigationScreenProp<NavigationRoute, Action>,
+    screenProps?: {},
+  ) => Options,
```

Comment:
Aha, good to see the `any` go away! This signature and name makes a lot of sense I think.